airbyte-integrations/connectors/source-microsoft-teams/source_microsoft_teams/client.py

```diff
@@ -0,0 +1,245 @@
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import csv
+import io
+import json
+import pkgutil
+from typing import Dict, Optional, Tuple
+
+import msal
+import requests
+from airbyte_protocol import AirbyteStream
+from msal.exceptions import MsalServiceError
+
+
+class Client:
+    MICROSOFT_GRAPH_BASE_API_URL = "https://graph.microsoft.com/"
+    MICROSOFT_GRAPH_API_VERSION = "v1.0"
+    PAGINATION_COUNT = 20
+
+    def __init__(self, config: object):
+        self.ENTITY_MAP = {
+            "users": self.get_users,
+            "groups": self.get_groups,
+            "group_members": self.get_group_members,
+            "group_owners": self.get_group_owners,
+            "channels": self.get_channels,
+            "channel_members": self.get_channel_members,
+            "channel_tabs": self.get_channel_tabs,
+            "conversations": self.get_conversations,
+            "conversation_threads": self.get_conversation_threads,
+            "conversation_posts": self.get_conversation_posts,
+            "team_drives": self.get_team_drives,
+            "team_device_usage_report": self.get_team_device_usage_report,
+        }
+        self.configs = config
+        self._group_ids = None
+        self.msal_app = msal.ConfidentialClientApplication(
+            self.configs["client_id"],
+            authority=f"https://login.microsoftonline.com/" f"{self.configs['tenant_id']}",
+            client_credential=self.configs["client_secret"],
+        )
+
+    def _get_api_url(self, endpoint: str) -> str:
+        api_url = f"{self.MICROSOFT_GRAPH_BASE_API_URL}{self.MICROSOFT_GRAPH_API_VERSION}/{endpoint}/"
+        return api_url
+
+    def _get_access_token(self) -> Dict[str, str]:
+        scope = ["https://graph.microsoft.com/.default"]
+        result = self.msal_app.acquire_token_silent(scope, account=None)
+        if not result:
+            result = self.msal_app.acquire_token_for_client(scopes=scope)
+        if "access_token" in result:
+            return result["access_token"]
+        else:
+            raise MsalServiceError(error=result.get("error"), error_description=result.get("error_description"))
+
+    def _make_request(self, api_url: str, params: Optional[Dict] = None):
```

Comment:
could you put return types on every method here? even if it ends up being `Dict`